zkevm-circuits/src/keccak_circuit/keccak_padding.rs

```diff
@@ -0,0 +1,427 @@
+
+            //TODO: could be removed if combined with keccak circuit?
+            for data_bit in d_bits {
+                let b = meta.query_advice(data_bit, Rotation::cur());
+                cb.require_boolean("input data bit", b);
+            }
+
+            for s_flag in s_flags {
+                let s = meta.query_advice(s_flag, Rotation::cur());
+                cb.require_boolean("boolean state bit", s);
+            }
+
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+
+                cb.require_boolean("boolean state switch", s_i - s_i_1);
+            }
+
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("padding bit checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut constraints = vec![];
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+                let d_bit_0 = meta.query_advice(d_bits[8 * i], Rotation::cur());
+                constraints.push(("begin with 1", (s_i - s_i_1) * (d_bit_0 - 1u64.expr())));
+            }
+            let s_last = meta.query_advice(s_flags[s_flags.len() - 1], Rotation::cur());
+            let d_last = meta.query_advice(d_bits[KECCAK_RATE - 1], Rotation::cur());
+
+            constraints.push(("end with 1", s_last * (d_last - 1u64.expr())));
+
+            cb.add_constraints(constraints);
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("intermedium 0 checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut sum_padding_bits = 0u64.expr();
+            for i in 0..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                sum_padding_bits = d_bits[i * 8..(i + 1) * 8]
+                    .iter()
+                    .map(|b| meta.query_advice(*b, Rotation::cur()))
+                    .fold(sum_padding_bits, |sum, b| sum + s_i.clone() * b);
+            }
+
+            cb.add_constraint("sum padding bits == 2", sum_padding_bits - 2u64.expr());
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("input len check", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut constraints = vec![];
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let len_i = meta.query_advice(d_lens[i], Rotation::cur());
+                let len_i_1 = meta.query_advice(d_lens[i - 1], Rotation::cur());
+                constraints.push(("len[i] = len[i-1] + !s_i", len_i - len_i_1 - not::expr(s_i)));
```

Comment:
Easier to read with `require_equal` (the code will almost be the same as the description :)